### PR TITLE
Improve user symbol resolution

### PR DESCRIPTION
**Pull request by lenticularis39**
This PR contains three related improvements of usym resolution (edit: added purpose of the changes):
- Caching BCC symcache per process if ASLR is enabled (edit: optional after changes, by default disabled)
  - Enables faster symbol resolution and symcache preloading (see below)
- Resolving symbols directly from ELF (only if ASLR is disabled)
  - Enables resolving symbols that are not dynamically loaded even after process exit
- Preloading BCC symcache for running instances of program
  - Enables resolving symbols after program exists even with ASLR enabled, given that the process exists at probe attachment time.

The latter two provide a partial fix of #2284. 

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by viktormalik** [comment]
CI failed, can you debug it?

**Comment by lenticularis39** [comment]
I looked into that, one issue is codegen tests, because there is now one more attribute of the usym event (that one is easily fixable), other issue is ustack runtime tests failing with segfault (that one will be a little bit more difficult, since it doesn't happen on my system).

Also there is another problem with the PR, that is that the necessary BCC fix was reverted in iovisor/bcc#4270 (it turned that `/proc/.../exe` is not a simple symlink, but it also reflects mount namespaces, meaning the patch broke symbol resultion for processes running in a different namespace). That has to be resolved in order for the last of this PR to work.

**Review by viktormalik — COMMENTED** [review]
Overall, these are very nice changes!

What I'm missing here, though, are better explanations of *why* the given approaches are necessary - this should appear in the PR description, in commit messages, and in code comments, too. Perhaps even a single-place summary of all the considered scenarios and corresponding solutions could be useful.

File: src/bpftrace.cpp
Comment (lenticularis39) [reply src/bpftrace.cpp:2190]: The issue with using insert is that under eager evaluation this always calls `get_symbol_table_for_elf`, making it less efficient.

Comment (viktormalik) [reply src/bpftrace.cpp:2200]: Ah, right. This looks ugly, but seems like we don't have much of a better option.
Perhaps this could be an alternative, but I'm not strongly against leaving the original one:
```suggestion
        if (symbol_table_cache_.find(pid_exe) == symbol_table_cache_.end())
            symbol_table_cache_[pid_exe] = get_symbol_table_for_elf(pid_exe)
        auto &symbol_table = symbol_table_cache_[pid_exe];
```

**Comment by lenticularis39** [comment]
Rebased and changed caching by PID, which is required for per-process symbol cache preloading, to optional (see the last commit for details).

File: tests/runtime/builtin
Comment (viktormalik) [inline tests/runtime/builtin:117]: For other tests, we use
```
REQUIRES bash -c "exit 1"
```
to explicitly disable the test.

Comment (lenticularis39) [reply tests/runtime/builtin:117]: Ah, nice, I was looking for a way to disable a test globally and I didn't find anything.

**Review by viktormalik — APPROVED** [review]
Looks good but it's quite a complex change so I'd like to have another review before merging.

File: man/adoc/bpftrace.adoc
Comment (ajor) [inline man/adoc/bpftrace.adoc:1770]:
```suggestion
If ASLR is enabled, user space symbolication only works when the process is running at either the time of the symbol resolution or the time of the probe attachment. The latter requires `BPFTRACE_CACHE_USER_SYMBOLS` to be set to `PER_PID`, and might not work with older versions of BCC. A similar limitation also applies to dynamically loaded symbols.
```
Nit: Reworded to sound better to me

File: src/bpftrace.cpp
Comment (ajor) [inline src/bpftrace.cpp:2203]: The old code had a matching `bcc_free_symcache` when caching was disabled - I assume we'll leak here without it

File: src/main.cpp
Comment (ajor) [inline src/main.cpp:312]:
```suggestion
    // bpftrace
```
The old name really refuses to go away! 😂

File: src/main.cpp
Comment (ajor) [inline src/main.cpp:335]: Can this just be an "else" or am I missing something?

File: man/adoc/bpftrace.adoc
Comment (ajor) [inline man/adoc/bpftrace.adoc:213]: This doesn't match the implemented behaviour, although the way it's described here sounds good to me

**Comment by ajor** [comment]
I love this PR, thanks so much for doing it - it solves so many problems I've had with user symbols over the years!

I'd suggest that we make "PER_PID" the default caching behaviour when ASLR is enabled, and leave it up to users to disable the cache if they feel it's using too much memory.
1. Having ASLR enabled is not too uncommon, and often users will only be tracing a single process. I'd think the large amount of forking described in #816 is a less common scenario
2. It's simpler for users to understand disabling a cache than it is to work out which type of caching to enable

File: src/bpftrace.cpp
Comment (ajor) [inline src/bpftrace.cpp:2158]: Is the caching just expected to fail when using wildcards in probes, e.g. "uprobe:/bin/gc*:*"

A comment explaining the expected behaviour would be useful for future developers

File: src/ast/passes/resource_analyser.cpp
Comment (ajor) [inline src/ast/passes/resource_analyser.cpp:241]: ustack is listed here, but it doesn't use the usym table, right?

File: src/required_resources.h
Comment (ajor) [inline src/required_resources.h:120]:
```suggestion
  std::unordered_set<std::reference_wrapper<Probe>> probes_using_usym;
```
Storing a reference will be a bit faster and use less memory than basing this set on strings. Theoretically we could also get probe name collisions, which references would prevent.

File: src/bpftrace.cpp
Comment (lenticularis39) [reply src/bpftrace.cpp:338]: It shouldn't be too much, e.g. `ls -l /proc/*/exe` takes 20ms on a system with around 350 processes, and the iteration only happens once per probe (which can be further reduced to just once with a better implementation).

File: src/ast/passes/resource_analyser.cpp
Comment (lenticularis39) [reply src/ast/passes/resource_analyser.cpp:241]: `ustack` uses `resolve_usym` to resolve addresses in the stack to symbols, as shown in your comment above, or did you mean something else?

File: man/adoc/bpftrace.adoc
Comment (lenticularis39) [reply man/adoc/bpftrace.adoc:213]: I don't see any difference between the behavior described here and the behavior implemented in `src/main.cpp`. Anyway, I'll change both to reflect the suggestion to default to PER_PID.

File: src/bpftrace.cpp
Comment (lenticularis39) [reply src/bpftrace.cpp:2158]: Yes, since the program is not expanded, we don't know which executable the perf event came from if a wildcard is used. I'll add a comment explaining this.

**Comment by lenticularis39** [comment]
Changed default setting of `BPFTRACE_CACHE_USER_SYMBOLS` to `PER_PROGRAM` without ASLR and `PER_PID` per ASLR, updated documentation, resolved comments, rebased to master and updated changelog with the new changes.

**Comment by ajor** [comment]
My comments about "ustack" come from it not passing a "probe_id" value back from the kernel like "usym" now does. The code is currently passing in "-1" as a probe_id for ustacks, which means they don't get all the features of usym (post-termination symbol resolution).

On another note: I'm wondering if we need to do some probe expansion when using usym now, i.e. similar to `probe_->need_expansion`.

I assume that would solve the wildcard issue I commented on before, and I'm also worried that we can potentially resolve symbols incorrectly against the wrong binary as it stands.

e.g. say we have a probe:
```probe:myprog1:funcA,uprobe:myprog2:funcB { @sym = usym) }```

If the process terminates before we resolve the symbol, I believe the current code will always pick "myprog1" as the pid_exe. If it was actually an address from myprog2, then we could end up with completely the wrong symbol. (not tested, just assuming)

Expansion would only really be needed when there are multiple different exes sharing a single probe. Multiple functions from the same exe sharing a probe would not need this. Multiple-function-single-exe is going to be a much more common case, and over-expanding probes would have some significant overheads when tracing a large number of functions, so is something we should avoid.

I realise that adding in this logic would not necessarily be trivial, and the rest of your changes in this PR look great, so if you want we could merge this with the post-process-termination logic temporarily disabled and follow up with a separate PR. Up to you.

**Comment by ajor** [comment]
Or maybe just explicitly disabling the post-process-termination logic for probes specifying multiple target paths (with either wildcards or commas) would be good enough as a permanent solution. That'd leave it working fast in the common case (probes with a single target path), and ensure we still get correct results, albeit slower, in the less common case.

**Comment by brendangregg** [comment]
I just heard about https://github.com/libbpf/blazesym which Meta is working on. At LSFMMBPF there was talk about whether it made sense for bcc and bpftrace to use blazesym (talk as in "does someone want to explore doing this?").

**Comment by lenticularis39** [comment]
I disabled ELF symbol resulution for probes with multiple attach points as suggested, since we don't know which binary is the correct one. `kstack` now passed `probe_id` to `resolve_usym`; I also updated the ELF symbol resolution logic to work with addresses with non-zero offsets, and removed the code for legacy BCC in `get_symbol_table_for_elf`, since we now ship BCC as a submodule.

(The failing tests are just a missing cast, I'll fix that soon.)

**Comment by viktormalik** [comment]
> I just heard about https://github.com/libbpf/blazesym which Meta is working on. At LSFMMBPF there was talk about whether it made sense for bcc and bpftrace to use blazesym (talk as in "does someone want to explore doing this?").

@brendangregg we were discussing this at one of the bpftrace office hours meetings. We're slowly trying to get rid of dependency on bcc and symbolization is one of those parts where we depend on it a lot. I'd keep that discussion outside of this PR, though.

File: src/bpftrace.h
Comment (ajor) [inline src/bpftrace.h:242]: I think uintptr_t would be a better type for storing symbols. It's what we use everywhere else at least

File: tests/runtime/builtin
Comment (ajor) [reply tests/runtime/builtin:119]: Ok, makes sense. I think there'd be a race condition between the target process exiting and bpftrace printing, but it'd be so extremely unlikely to happen that it doesn't matter for this.

What's the race condition you mention in the comment about this test being disabled though?

**Review by ajor — COMMENTED** [review]
> I also updated the ELF symbol resolution logic to work with addresses with non-zero offsets

Did you have an example of where this was causing symbolication problems? It'd be useful to check that in as a test-case.

I have a slight concern about the impact this pre-caching will have on startup time when tracing huge programs. I'll do a little bit of testing on this to satisfy myself.

File: src/utils.cpp
Comment (lenticularis39) [reply src/utils.cpp:1340]: That is because the semantics of `std::map::lower_bound` and `std::map::upper_bound`, citing CPPreference:
- `std::map::lower_bound`: Returns an iterator pointing to the first element that is not less than (i.e. greater or equal to) key.
- `std::map::upper_bound`:  Returns an iterator pointing to the first element that is greater than key.

I want the element that is *less or equal* to the key (possible start address of the symbol), therefore I have to sort the elements of the map the other way.

File: tests/runtime/builtin
Comment (lenticularis39) [reply tests/runtime/builtin:119]: Just for clarification, the sequence of events that happens 99%+ of the time:
- `uprobe_symres_exited_process` starts and sleeps for 3 seconds.
- bpftrace attaches probe on `uprobe_symres_exited_process:test` and pre-loads the cache with the use of util function `get_pids_for_program` and `bcc_symcache_new`.
- the process wakes up, executes `test()` and immediately exists
- the probe fires and `bcc_symcache_resolve` is run on the stored cache (note that `symbol_table_cache_` is not used here, since `BPFTRACE_CACHE_USER_SYMBOLS` is set to `PER_PID`)

Now what happens in `bcc_symcache_resolve` is that the binary of `uprobe_symres_exited_process` is opened to look for the symbol. However, since in a general case the binary can be in a different mount namespace, BCC uses a file descriptor of `/proc/<pid>/root` that was pre-opened in `bcc_symcache_now` to access the file using `openat` (see iovisor/bcc#4319). The issue is that this approach with `openat` sometimes does not work due to current implementation issues, hence BCC checks if the process is running, and if yes, it does not use `openat`. This is prone to a race condition, since the process may exit *after* this check is done.

This race condition is much rarer than the original one from #2284 linked above (which in my case happened 100% of the time), but still happens and would cause CI randomly failing, hence the test being disabled.

**Comment by lenticularis39** [comment]
> > I also updated the ELF symbol resolution logic to work with addresses with non-zero offsets
> 
> Did you have an example of where this was causing symbolication problems? It'd be useful to check that in as a test-case.

The issue is in general `ustack` addresses that are not resolvable using `bcc_symcache_resolve`, but are resolvable using `symbol_table_cache_`. More specifically, it is `ustack` addresses in exited processes with ASLR disabled - I didn't write any tests for this because of the requirement of disabling ASLR (also for `usym` resolution), now when I'm thinking about it I could use `personality(ADDR_NO_RANDOMIZE)` to disable ASLR in a test program and test both.

> I have a slight concern about the impact this pre-caching will have on startup time when tracing huge programs. I'll do a little bit of testing on this to satisfy myself.

OK. The bcc symcache pre-loading is not a issue, since it only loads the symbols when `bcc_symcache_resolve` is called (since e.g. dynamically loaded modules might not be present at the time of calling `bcc_symcache_new`), what could impact the startup time is `symbol_table_cache_`, it might be useful to have an option to disable it, like we do for the symcache.

File: src/types.cpp
Comment (lenticularis39) [reply src/types.cpp:122]: Wouldn't that then also be a case for `ustack`, which is already defined as a byte array and realized with an 8-aligned struct in codegen?

File: src/bpftrace.h
Comment (lenticularis39) [reply src/bpftrace.h:242]: Makes sense, on 32-bit platforms it will save some memory.

File: src/ast/passes/codegen_llvm.cpp
Comment (lenticularis39) [reply src/ast/passes/codegen_llvm.cpp:154]: That shouldn't be too hard, I'll try to add it with improvements from the other comments, if it turns out to be too difficult to quickly implement, I'll add a comment.

File: src/types.cpp
Comment (ajor) [reply src/types.cpp:122]: Good point, and I don't know...

@viktormalik Can you shed any light on this?

File: tests/testprogs/disable_aslr.c
Comment (lenticularis39) [reply tests/testprogs/disable_aslr.c:7]: I am wondering if this can be solved by an existing program, I don't know any.

**Comment by lenticularis39** [comment]
Oops, I forgot to change the main offset in the test to a more generic pattern, causing the test to fail with a different compiler.

File: tests/testprogs/disable_aslr.c
Comment (viktormalik) [reply tests/testprogs/disable_aslr.c:7]: Since it's in tests, there's no need to fix this. Dismissed.

File: src/types.cpp
Comment (viktormalik) [reply src/types.cpp:122]: So, it turns out that the `SizedType::GetAlignment` function is only used to compute alignment in tuples. I tried to change it to 8 but that causes some failures, so I'll need to investigate more.

Comment (viktormalik) [reply src/types.cpp:122]: This is now addressed in #2625.

**Review by ajor — APPROVED** [review]
Looks great, thank you!

**Comment by lenticularis39** [comment]
Rebased and fixed failing test added by #2625.

**Pull request merged**
